File: BinarySearchTree.py

class Node:

    left_node = None
    right_node = None
    data = None

    def __init__(self):
        self.data=None

    def get_data(self):
        return self.data

    def set_data(self,data):
        self.data=data
    
    def get_right_node(self):
        return self.right_node

    def get_left_node(self):
        return self.left_node

    def set_right_node(self, right_node):
        self.right_node = right_node

    def set_left_node(self, left_node):
        self.left_node = left_node

    def __str__(self):
        return str(self.data)

    def __iter__(self):
        return inOrderTraversal(self)





class BinarySearchTree:

    name = None
    root = None

    def __init__(self, name, root):
        self.name = name
        self.root=root

    def set_root(self,root):
        self.root = root

    def get_root(self):
        return self.root

    def add_all(self, *list):
        self.root=None
        self.root=Node()
        self.root.set_data(list[0])
        for x in range (1,len(list)):
            self.add_element(list[x],self.root,Node())


    def add_element(self, element ,currentNode, node_to_be_added):

        if currentNode.get_data() > element:
            if currentNode.get_left_node() is None:
                node_to_be_added.set_data(element)
                currentNode.set_left_node(node_to_be_added)
            else:
                self.add_element(element,currentNode.get_left_node(),node_to_be_added)

        else:
            if currentNode.get_right_node() is None:
                node_to_be_added.set_data(element)
                currentNode.set_right_node(node_to_be_added)
            else:
                self.add_element(element,currentNode.get_right_node(),node_to_be_added)


    def __str__(self):
        return "["+self.name+"]"+self.preOrderTraversal("",self.root)



    def preOrderTraversal(self,str,currentNode):
        str=str+currentNode.__str__()
        if currentNode.get_left_node() is not None:
            str=str+" L:("
            str=self.preOrderTraversal(str,currentNode.get_left_node())+")"


        if currentNode.get_right_node() is not None:
            str=str+" R:("
            str=self.preOrderTraversal(str,currentNode.get_right_node())+")"

        return str

    """
    def inorderTraversal(self,currentNode,list):
        if currentNode.get_left_node() is not None:
            list= self.inorderTraversal(currentNode.get_left_node(),list)

        list.append(currentNode)

        if currentNode.get_right_node() is not None:
            list=self.inorderTraversal(currentNode.get_right_node(),list)

        return list
    """
"""
def inorder(t):


    if t:
        for x in inorder(t.left_node):
            yield x

        yield t.data

        for x in inorder(t.right_node):
            yield x
            """




def inOrderTraversal(currentNode):

    if currentNode.get_left_node() is not None:
        for d in currentNode.get_left_node().__iter__():
            yield d

    yield currentNode

    if currentNode.get_right_node() is not None:
        for d in currentNode.get_right_node().__iter__():
            yield d

class Merger:

    def merge(t1,t2):
        merged_list=[]

        list1 = []
        list2 = []

        for x in t1.root:
            list1.append(x)

        for x in t2.root:
            list2.append(x)

        index1 = 0
        index2 = 0

        remaining_index = 0

        while index1<len(list1) and index2<len(list2):
            if(list1[index1].get_data()<list2[index2].get_data()):
                merged_list.append(list1[index1].get_data())
                index1=index1+1

            else:
                merged_list.append(list2[index2].get_data())
                index2=index2+1


        if(index1==len(list1)):
            remaining_index=len(list2)-index2
            for i in range(0,remaining_index):
                merged_list.append(list2[index2].get_data())
                index2 = index2 + 1

        else:
            remaining_index = len(list1) - index1
            for i in range(0, remaining_index):
                merged_list.append(list1[index1].get_data())
                index1= index1 + 1

        return merged_list


if __name__ == "__main__":
    t1 = BinarySearchTree(name="Oak", root=Node())
    t2 = BinarySearchTree(name="Birch", root=Node())
    t3= BinarySearchTree(name="One Node Tree",root=Node())

    t1.add_all(5, 3, 9, 0)
    t2.add_all(1, 0, 10, 2, 3, 6, 5, 7, 8, 4, 9)
    t3.add_all(20)
    print(t1)
    print(t2)
    print(t3)

    for x in t1.root:
        print(x)
    for x in t2.root:
        print(x)
    for x in t3.root:
        print(x)
    print(Merger.merge(t1, t2))

    tree_one = BinarySearchTree(name="merge1", root=Node())
    tree_two = BinarySearchTree(name="merge2", root=Node())

    tree_one.add_all(14, 11, 8, 5, 6, 9, 12, 15, 18)
    tree_two.add_all(10, 9, 16, 13, 17)

    print(tree_one)
    print(tree_two)
    print(Merger.merge(tree_one,tree_two))

File: BinarySearchTree.java
import java.util.*;
import java.util.function.Consumer;

class Node<T extends Comparable>  {


    T data;
    Node left;
    Node right;

    boolean isVisited;

    Node()
    {
        isVisited=false;
    }

    Node(T t)
    {
        data=t;
        isVisited=false;
    }

    public void setLeft(Node left) {
        this.left = left;
    }

    public void setRight(Node right) {
        this.right = right;
    }

    public void setData(T data) {
        this.data = data;
    }

    public Node getLeft() {
        return left;
    }

    public Node getRight() {
        return right;
    }

    public T getData() {
        return data;
    }

    public void setVisited(boolean visited) {
        isVisited = visited;
    }

    public boolean isVisited() {
        return isVisited;
    }

    @Override
    public String toString() {
        return data.toString();
    }
}

public class BinarySearchTree<T extends Comparable> implements Iterable{

    String name;

    Node root;

    public Node getRoot() {
        return root;
    }

    int numberOfNodes;

    public int getNumberOfNodes()
    {
        return numberOfNodes;
    }
    BinarySearchTree(String name)
    {
        this.name=name;
    }

    BinarySearchTree()
    {

    }

    public void addAll(List<T> toBeAdded)
    {
        this.numberOfNodes=toBeAdded.size();

        if(toBeAdded!=null||toBeAdded.size()>0) {
             root = new Node( toBeAdded.get(0));
             listOfNodes.add(root);



            for(int i=1; i<toBeAdded.size();i++)
            {
               iterateOverTree(root,toBeAdded.get(i));
            }
        }
    }

    List<Node<T>> listOfNodes=new ArrayList<>();

    public  void iterateOverTree(Node currentNode, T data)
    {
        Node node=new Node(data);

        if(data.compareTo(currentNode.getData())<0)
        {
            if(currentNode.getLeft()!=null)
                iterateOverTree(currentNode.getLeft(), data);

            else {
                currentNode.setLeft(node);
                listOfNodes.add(node);
            }
        }
        else
        {
            if(currentNode.getRight()!=null)
                iterateOverTree(currentNode.getRight(), data);

            else {
                currentNode.setRight(node);
                listOfNodes.add(node);
            }
        }
    }

    int nextCounter=0;
    @Override
    public Iterator iterator() {

        Node start=this.root;

        BinarySearchTree<T> temp=this;

    Iterator<BinarySearchTree<T>> newIterator= new Iterator<BinarySearchTree<T>>() {
        @Override
        public boolean hasNext() {
            if(nextCounter< temp.getNumberOfNodes())
                return true;

            else return false;
        }

        @Override
        public BinarySearchTree<T> next() {

            try {
                nextCounter++;
                inorderTraversal(temp.root);
            } catch (IllegalArgumentException e) {

            } finally {


                Node<T> newNode = reference;

                BinarySearchTree<T> toBeReturned = new BinarySearchTree<>();

                toBeReturned.addAll(Arrays.asList(newNode.getData()));

                if (nextCounter == temp.numberOfNodes - 1) {
                    reference = null;
                }

                return toBeReturned;
            }
        }


    };


    return newIterator;

    }

    @Override
    public void forEach(Consumer action) {

        Iterator newIterator=this.iterator();

        while(newIterator.hasNext())
        {
            BinarySearchTree<T> binarySearchTree= (BinarySearchTree<T>) newIterator.next();

            System.out.println(binarySearchTree.root.toString());

        }
        nextCounter=0;

        this.resetSetVisited();
        /*for(int i=0;i<this.listOfNodes.size();i++)
        {
            listOfNodes.get(i).setVisited(false);
        }*/
    }

    @Override
    public String toString() {
        return printAll(this.root,"["+this.name.toString()+"]"+" ");
    }

    private String printAll(Node currentRoot,String s) {

        if(currentRoot.getLeft()!=null)
        s=s+currentRoot.getData().toString()+" ";

        else
            s=s+currentRoot.getData().toString()+"";


       if(currentRoot.getLeft()!=null)
        s=printAll(currentRoot.getLeft(),s+"L:(")+")";

        if(currentRoot.getRight()!=null) {
            s = printAll(currentRoot.getRight(), s + " R:(")+")";
        }

        return s;

    }

    Node<T> reference=new Node<>();

    public Node<T> getReference() {
        return reference;
    }

    public void inorderTraversal(Node currentNode) throws IllegalArgumentException
    {
       if(currentNode.left!=null)
           inorderTraversal(currentNode.left);

       if(currentNode.isVisited())
       {
           if(currentNode.right!=null)
               inorderTraversal(currentNode.right);
       }

        else{
            currentNode.setVisited(true);
            reference=currentNode;
            throw new IllegalArgumentException();

        }

    }


    public static <T extends Comparable<T>> List<T> merge(BinarySearchTree<T> t1, BinarySearchTree<T>t2)
    {

        List<T> listToBeAdded=new ArrayList<>();

       int firstTreeNodeCounter=0;
       int secondTreeNodeCounter=0;

        Thread firstThread=new Thread(new Runnable() {
           @Override
           public void run() {

               try {
                   t1.inorderTraversal(t1.getRoot());
               }
               catch (IllegalArgumentException e)
               {

               }
           }
       });

       Thread secondThread=new Thread(new Runnable() {
           @Override
           public void run() {

               try {
                   t2.inorderTraversal(t2.getRoot());
               }
               catch(IllegalArgumentException e)
               {

               }
           }
       });



       boolean firstTreeStopped=true;
       boolean secondTreeStopped=true;


       while(firstTreeNodeCounter<=t1.numberOfNodes-1&&secondTreeNodeCounter<=t2.numberOfNodes-1)
       {
          if(firstTreeStopped)
              firstThread.run();

          if(secondTreeStopped)
              secondThread.run();

          if(t1.getReference().getData().compareTo(t2.getReference().getData())<=0)
           {
               if(!firstTreeStopped)
                   firstTreeStopped=true;


              listToBeAdded.add(t1.reference.getData());
              firstTreeNodeCounter++;
              secondTreeStopped=false;
           }

           else
           {
               if(!secondTreeStopped)
                   secondTreeStopped=true;

               listToBeAdded.add(t2.reference.getData());
               secondTreeNodeCounter++;
               firstTreeStopped=false;
           }
       }
       int difference=0;

       if(firstTreeNodeCounter<t1.getNumberOfNodes()) {
           difference = t1.getNumberOfNodes() - firstTreeNodeCounter-1;
           listToBeAdded.add(t1.getReference().getData());
           firstTreeStopped=true;
           secondTreeStopped=false;
       }

       if(secondTreeNodeCounter<t2.getNumberOfNodes()) {
           difference = t2.getNumberOfNodes() - secondTreeNodeCounter-1;
           listToBeAdded.add(t2.getReference().getData());
           firstTreeStopped = false;
           secondTreeStopped = true;
       }

       for(int i=0;i<difference;i++)
       {
           if(firstTreeStopped&&!secondTreeStopped) {
               firstThread.run();
               listToBeAdded.add(t1.reference.getData());
           }

           else {
               secondThread.run();
               listToBeAdded.add(t2.reference.getData());
           }

       }

       t1.resetSetVisited();
       t2.resetSetVisited();

       return listToBeAdded;
    }

    public void resetSetVisited()
    {
        for(int i=0;i<this.listOfNodes.size();i++)
        {
            listOfNodes.get(i).setVisited(false);
        }
    }

    public static void main(String[] args) {
// each tree has a name, provided to its constructor
        BinarySearchTree<Integer> t1 = new BinarySearchTree<>("Oak");
// adds the elements to t1 in the order 5, 3, 0, and then 9
        t1.addAll(Arrays.asList(5, 3, 0, 9));
        BinarySearchTree<Integer> t2 = new BinarySearchTree<>("Maple");
// adds the elements to t2 in the order 9, 5, and then 10
        t2.addAll(Arrays.asList(9, 5, 10));
        System.out.println(t1);// see the expected output for exact format
        t1.forEach(System.out::println);// iteration in increasing order
        System.out.println(t2); // see the expected output for exact format
        t2.forEach(System.out::println); // iteration in increasing order
        BinarySearchTree<String> t3 = new BinarySearchTree<>("Cornucopia");
        t3.addAll(Arrays.asList("coconut", "apple", "banana", "plum", "durian",
                "no durians on this tree!", "tamarind"));
        System.out.println(t3); // see the expected output for exact format
        t3.forEach(System.out::println);

        BinarySearchTree<Double> t4 = new BinarySearchTree<>("Double");
        t4.addAll(Arrays.asList(3.2,6.3,9.4,12.2));
        System.out.println(t4); // see the expected output for exact format
        t4.forEach(System.out::println); // iteration in increasing order

        List<Integer> integerList=merge(t1,t2);

        for(int i=0; i<integerList.size();i++)
        {
            System.out.println(integerList.get(i));
        }
    }


}
